servers/src/main/java/tachyon/worker/file/FileDataManager.java

```diff
@@ -75,13 +77,21 @@ public synchronized void persistFile(long fileId, List<Long> blockIds) throws IO
     OutputStream outputStream = mUfs.create(dstPath);
     final WritableByteChannel outputChannel = Channels.newChannel(outputStream);
 
+    Map<Long, Long> blockIdToLockId = Maps.newHashMap();
+    // lock all the blocks to prevent any eviction
     for (long blockId : blockIds) {
       long lockId;
       try {
         lockId = mBlockDataManager.lockBlock(Sessions.CHECKPOINT_SESSION_ID, blockId);
+        blockIdToLockId.put(blockId, lockId);
       } catch (BlockDoesNotExistException e) {
         throw new IOException(e);
       }
+    }
+
+    for (int i = 0; i < blockIds.size(); i ++) {
+      long blockId = blockIds.get(i);
+      long lockId = blockIdToLockId.get(blockId);
 
       // obtain block reader
       try {
```

Comment:
There are several nested try{} blocks. A simpler way might be to have a single try{}, put the for loop within the try, catch any types of Exceptions, and only in the finally{} block, release all the locks.